### ISSUE TITLE
gccxml fatal error and stalled vsyncs

### ISSUE THREAD

**Issue by xposure** [sharpdx/SharpDX#582]
TLDR
*  SwapChain.Present is missing vsyncs giving me 30 and 15fps with less than 5% gpu load and less than 3% cpu load (empty game instances). I have latest drivers for nvidia, tried disabling SLI, enabling SLI, all forms of adaptive and smooth sync using nvidia inspector and can get rid of the issue. 
*  Can not get SharpDX to build with the follow error `in gccxml Unable to locate key [SOFTWARE\Microsoft\Windows Kits\Installed Roots] in registry`. I installed, dx june sdk, win7.1 and 8.1 sdk, i have vs2010, vs2012, vs2013 installed, i do not have KitRoot10 (windows 10 sdk perhaps?) registry entry and running windows 7 trying to build master branch.

Long Version

I've been working on trying to figure out for weeks why the SwapChain.Present is taking 33ms to 66ms when vsync is enabled and I launch multiple instances of an empty project in MonoGame (or simply move my mouse around over chrome or have chrome play a video - all directx).

Issue 1: https://github.com/mono/MonoGame/issues/3553
Issue 2: https://github.com/mono/MonoGame/issues/3660
Possibly related: https://github.com/sharpdx/SharpDX/issues/579

To summarize the issues, if I have multiple instances of a completely empty monogame project (using sharpdx 2.6.3 I believe) or I have just one and I use chrome and leave a input field focused or a youtube video playing, SwapChain.Present misses vsyncs and will start rendering at 30 and 15 fps. I have no FPS issues or stuttering with any games on this machine (I own quite a few on steam).

So to eliminate that MonoGame is the culprit (they haven't been very helpful with this issue), I decided to get SharpDX and run a barebones project and see if I can reproduce it to determine if MonoGame is setting up SharpDX wrong or if there is an issue with drivers fighting a race condition.

But one does not simply get SharpDX. 

*  You could checkout the SharpDX-Samples and follow the outdated instructions and cuss out nuget for an hour or search and realize that the samples rely on SharpDX source and submodules now. 
*  You could perhaps get the binaries off the site, but those don't include the Desktop DLL or the Mathematics DLL to run the samples provided. 
*  You could then try to get the SharpDX source and follow several different outdated instructions for building the source. Install vs2010, install vs2012, install windows 7.1 sdk, install windows 8.1 sdk, dx june sdk and still have gccxml tell you it can't find a registry key (which doesn't exist at all on my machine) HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows Kits\Installed Roots\KitsRoot10
*  You could continue to beat your head against a wall or ask for help - this is the step I am on.

**Comment by xoofx** [comment]
Sorry for the current state of the repository for building it. I had no spare time recently to cleanup the doc and readme, this will be improved in the coming weeks.

Though, whenever you need to checkout a specific version, you can always do it by using the appropriate tag and branch from it (in your case [2.6.3](https://github.com/sharpdx/SharpDX/tree/2.6.3))

Concerning your specific issue, you can also download the 3.0.0-alpha01 binaries from sharpdx.org (Latest Dev Package button from Home page), unzip this in a SharpDX directory, and clone this Samples directory in it. All the samples should run from there.

I have tested this with the MultiCube Direct3D11 sample which is using Direct3D11 + Deferred context. Edit the code where there is a Present and add a Present(1,...) instead of Preset(0, ...). Run this application multiple times on your machine. I just tested with 3 instances and FPS is constant at 60 FPS in each app/window (unless you move them).

The code behind the Present in SharpDX is a direct call to DXGI, so It is unlikely a bug in SharpDX but a setup/usage/driver problem (in the code and/or in your computer)

**Comment by cphillips83** [comment]
@xoofx I agree that there is something going on at a much lower level than SharpDX and I'll try your instructions here today and see what my results are.

The problem I face tho is that I don't experience this with any other games on this system and I have experienced it with 2 out of 3 computers at the moment which makes me very hesitant to use SharpDX. I need to figure out exactly where this issue is at before I can move further with my project or re-evaluate backends.


Are there any low levels tools that could give me detailed diag on what exactly DirectX is doing?

**Comment by xoofx** [comment]
First you should check in debug mode that there is no Direct3D message warnings that could indicate a wrong usage of the API (with Direct3D11 Device created in debug, unsafe debugging tun-on on your csproj in VS and checkout the VS Output pane)

Then there is tools like GPUView http://graphics.stanford.edu/~mdfisher/GPUView.html

**Comment by xposure** [comment]
@xoofx Well heres my update, the good news is I can reproduce it with MultiCube D3D example. It runs at about 7000 fps with one instance and no vsync and about 800 fps with 6 running. The bad news is I still don't have a clue.

When I enable vsync and run 4 of them, they all dropped to 15FPS so after some head aches I got GPUView up and running but this thing is a beast and a bit over my head at the moment. I also tried to do unsafe debugging but ran in to a lot of issues. Best I could come up with was you either need to switch the projects to framework 4 or higher, switch to x86 and there was another I forget. I switched to x86 in the config and nothing new came up in the debug output other than a few ms bugs.

Here is a visual image of GPUView and it showing the VSyncs being skipped in a linear fashion. I really feel that this is something to do with SLI and NVidia's adaptive vsync but I tried all settings and it didn't appear to make a difference (using nvidia inspector). On the other hand, its acting like some sort of mutex is being locked and forcing each app to miss it and then the next app in line grabs it as soon as the first one let it go and so on creating a nice round robin race condition.

Any suggestions for me to dig in to this further or things to try to pinpoint this?

![vsync](https://cloud.githubusercontent.com/assets/6292318/7263138/fcfbf49c-e84c-11e4-8d55-cd8dcb3d2e28.png)

**Comment by xoofx** [comment]
Well, you should try to reproduce this with a small native C++ program and go to a NVidia forum to report this. I don't think that it is related to SharpDX, unless proven otherwise.

**Comment by xposure** [comment]
I guess I'll give that a try. After more work and research its definitely something down in the nvidia driver. I was able to reproduce this with a real game using DX11 (Elite Dangerous). Thanks for the help an the introduction to GPUView.
